File: game_agent.py
import argparse, json
import util



def game_agent(filename, enum, board_init, random_seed):
    game = util.yaml2bt(filename, True, True)
    engine, max_tile_width, rng = util.setup_engine(game, 'UNDO_NONE', board_init, random_seed)

    queue = []
    seen = {}
    enum_seen = {}

    engine.stepToWaitChoiceOrResult()
    state = engine.getState()
    stateStr = str(state)

    queue.append((state, 0))
    seen[stateStr] = None

    while len(queue) > 0:
        state, steps = queue[0]
        queue = queue[1:]

        engine.setState(state)
        stateBoard = dict(state.board)
        stateResult = dict(state.gameResult) if engine.gameOver() else None

        if enum:
            enum_key = str((stateBoard, stateResult))
            if enum_key not in enum_seen:
                enum_seen[enum_key] = None
                print(json.dumps({'board':stateBoard, 'steps':steps, 'game_result':stateResult}), flush=True)

        if stateResult is not None:
            if stateResult['result'] == 'win':
                if not enum:
                    print(json.dumps({'board':stateBoard, 'steps':steps, 'game_result':stateResult, 'success':True}), flush=True)
                    return
            else:
                continue

        else:
            if state.displayWait:
                raise RuntimeError('At display wait.')
            if not state.choiceWait:
                raise RuntimeError('Not at choice wait.')

            for choiceIndex in range(len(state.choices)):
                engine.setState(state)
                engine.clearChoiceWait(True, choiceIndex)
                engine.stepToWaitChoiceOrResult()
                nextState = engine.getState()
                nextStateStr = str(nextState)
                if nextStateStr not in seen:
                    queue.append((nextState, steps + 1))
                    seen[nextStateStr] = None

    if not enum:
        print(json.dumps({'success':False}), flush=True)



if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Run agent on game YAML.')
    parser.add_argument('filename', type=str, help='Filename to process.')
    parser.add_argument('--enum', action='store_true', help='Enumerate boards rather than find winning board.')
    parser.add_argument('--board', type=str, help='Initial board configuration.')
    parser.add_argument('--seed', type=int, help='Random seed.')
    args = parser.parse_args()

    board_init = None if args.board is None else json.loads(args.board)

    game_agent(args.filename, args.enum, board_init, args.seed)

File: script/game_agent.js
const args = parse_args({
    game: { type: 'string', required: true },
    board: { type: 'string' },
    enumerate: { type: 'boolean', default: false },
});

const game = load_game(args.game, true, true);

const board_init = args.board ? JSON.parse(args.board) : null;

const [engine, max_tile_width] = setup_engine(game, ENG_UNDO_NONE, board_init, 0);

let queue = [];
let seen = {};
let enum_seen = {};

engine.stepToWaitChoiceOrResult();
const stateInit = engine.getState();
const stateInitStr = JSON.stringify(stateInit);

queue.push([stateInit, 0]);
seen[stateInitStr] = null;

while (queue.length > 0) {
    const [state, steps] = queue.shift();

    if (args.enumerate) {
        const enum_key = JSON.stringify([state.board, state.gameResult]);
        if (!(enum_key in enum_seen)) {
            enum_seen[enum_key] = null;
            process.stdout.write(JSON.stringify({'board':state.board, 'steps':steps, 'game_result':state.gameResult}) + '\n')
        }
    }

    if (state.gameResult !== null) {
        if (state.gameResult.result === 'win') {
            if (!args.enumerate) {
                process.stdout.write(JSON.stringify({'board':state.board, 'steps':steps, 'game_result':state.gameResult, 'success':true}) + '\n');
                process.exit();
            } else {
                continue;
            }
        }
    } else {
        if (state.displayWait) {
            throw new Error('At display wait.');
        }
        if (!state.choiceWait) {
            throw new Error('Not at choice wait.');
        }

        for (const choiceIndex in state.choices) {
            engine.setState(state);
            engine.clearChoiceWait(true, choiceIndex);
            engine.stepToWaitChoiceOrResult();
            const nextState = engine.getState();
            const nextStateStr = JSON.stringify(nextState);
            if (!(nextStateStr in seen)) {
                queue.push([nextState, steps + 1]);
                seen[nextStateStr] = null;
            }
        }
    }
}

if (!args.enumerate) {
    process.stdout.write(JSON.stringify({'success':false}) + '\n');
}
